LayoutTests/http/tests/push-api/subscribe-deny-permissions-on-prompt.html

```diff
@@ -16,35 +16,30 @@
         await testServiceWorkerSubscribe(registration, 'NotAllowedError');
         await testDocumentSubscribeWithoutUserGesture(registration, 'NotAllowedError');
 
-        let promise = new Promise((resolve, reject) => {
-            if (!window.internals) {
-                reject('Test requires internals.');
-                return;
-            }
+        if (!window.internals)
```

Comment:
Can you explain in commit message that why this change is needed? It's not clear to me why pushManager would return different results after this patch

Reply:
`WebNotificationClient::checkPermission` (and thus, `Notification::permission`) used to get the permission state from a cache in `WebNotificationManager`. This cache did not get updated by calling `TestController::denyNotificationPermissionOnPrompt` and then `WebNotificationClient::requestPermission` (which is what happens in this test). So when request2 occurred, `WebNotificationClient::checkPermission` would return default, then fail the transient activation check and reject the promise with a "user gesture" error. 

This patch removes this cache and refactors `WebNotificationClient::checkPermission` (and thus, `Notification::permission`) to get the permission state from the client (which is `TestController` in the case of this test). The cache in `TestController` that stores the permissions state does get updated by calling `TestController::denyNotificationPermissionOnPrompt` and then `WebNotificationClient::requestPermission`. So now, when request2 occurs, `WebNotificationClient::checkPermission` returns denied and rejects the promise with a "user denied push permission" error before reaching the transient activation check.

I spoke with @bnham about this and learned that the point of this layout test is to check that given a user gesture, the first call to `PushManager::subscribe` would consume the transient activation and that the second call would not have a transient activation. Due to the change described above, this check on the second call was being circumvented. We can ensure that this check occurs by altering the test the way I did in this patch.

You're right, the reason for change is not clear just by looking at the test, so I'll update the commit message with the above explanation once the bots finish and I make any other suggested changes.

Reply:
I believe WebKitTestRunner already caches the result. `TestController::decidePolicyForNotificationPermissionRequest.` already calls `setPermission` on `WebNotificationProvider`. denyOnPrompt means that if you request/prompt for permission, the state will be set to denied. The test was expecting that after denyOnPrompt is called, then first request does a prompt, the second request will get that the permission state is default? But that doesn't seem to make sense. The permission state should be denied, no? The reason that the test was passing before this patch was because it was relying on the fact that WebKitTestRunner wasn't updating the cache in WebNotificationManager. Now that this cache is gone, and checkPermission actually goes to the client, and testRunner returns the result based on its cache in WebNotificationProvider, the second request correctly gets back that the permission state is denied.